activerecord/lib/active_record/relation/predicate_builder.rb

```diff
@@ -145,5 +150,9 @@ def can_be_bound?(column_name, value)
         handler_for(value).is_a?(BasicObjectHandler) &&
         !table.associated_with?(column_name)
     end
+
+    def polymorphic_array?(value)
+      value.is_a?(Array) && value.map(&:class).map { |e| e.try(:base_class) }.uniq.size > 1
```

Comment:
@sgrif Sorry, for my unfamiliarity, but at this point `value` would be an array of AR objects, not necessarily a relation, so I don't think I would have access to `polymorphic?`  If that's the case, I'm going to try your suggestion of moving the switch to a factory method on AssociationQueryValue and move this there with an early exit.